lib/elixir/lib/string_io.ex

```diff
@@ -263,37 +257,21 @@ defmodule StringIO do
 
   ## get_line
 
-  defp get_line(encoding, prompt,
-                %{input: input, output: output, capture_prompt: capture_prompt} = s) do
-    case :unicode.characters_to_list(input, encoding) do
-      {:error, _, _} ->
-        {{:error, :collect_line}, s}
-      {:incomplete, _, _} ->
-        {{:error, :collect_line}, s}
-      chars ->
-        {result, input} = do_get_line(chars, encoding)
+  defp get_line(encoding, prompt, %{input: input} = s) do
+    with {result, input} <- binary_collect(input, encoding, "") do
+      result = case result do
+        "" -> :eof
+        r -> r
+      end
 
-        s =
-          if capture_prompt do
-            %{s | output: <<output::binary, IO.chardata_to_string(prompt)::binary>>}
-          else
-            s
-          end
+      s = capture_prompt(s, prompt)
 
-        {result, %{s | input: input}}
+      {result, %{s | input: input}}
```

Comment:
First, let me make sure I understand you correctly. Are you suggesting replacing `capture_prompt/2` with a function like:

```elixir
defp state_after_read(%{capture_prompt: false} = s, remainder, _prompt) do
  %{s | input: remainder}
end
defp state_after_read(%{capture_prompt: true, output: output} = s, remainder, prompt) do
  %{s | input: remainder, output: <<output::binary, IO.chardata_to_string(prompt)::binary>>}
end
```

so that 

```
s = capture_prompt(s, prompt)
{result, %{s | input: remainder}}
```

can be replaced by 

```elixir
{result, state_after_read(s, remainder, prompt)}
```

?

This works but has no measurable impact on performance (setting an item in a map is fast). It's a question of code style and I don't have a preference. Please let me know if you think.